### PR TITLE
Update script composable update method draft

### PR DESCRIPTION
**Pull request by mzgoddard**
#### Summary

Draft an implementation of the mdn-bcd-collector update script with its `update` method written as a composition of smaller functions.

#### Test results and supporting details

This change is incomplete and untested. It is meant to check that this is a useful change to make to facilitate further work on the script.

I've structured this version of the `update` function as a set of smaller functions that operate on an `UpdateState` object. The functions either provide set fields in the state or filter the state and give a reason. Some of the functions providing state expand it to iterate over substates, like browsers.

One of my goals for this change is to produced machine readable logs from it. Likely an array of objects containing an UpdateState (or portion of it), the action taken (filter or persist), and reason for the action.

I've also included a version of `inferSupportStatements`. It works in two loops, first to collapse ranges of null supported versions with their bookend versions, and the second turning that into `SimpleSupportStatement[]`.

#### Related issues

https://github.com/mdn/browser-compat-data/issues/19868

### CODE REVIEW

File: scripts/update-2.ts [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    >) => {
+      if (typeof inferredStatement.version_removed === 'string') {
+        simpleStatement.version_removed = inferredStatement.version_removed;
+        return state({ statements: allStatements });
+      }
+      return undefined;
+    },
+  );
+
+const update = (bcd, supportMatrix, filter) => {
```
Comment (mzgoddard) [inline scripts/update-2.ts:538]: Starting from here may help.

**Comment by jugglinmike** [comment]
This seems potentially useful, though it'd be easier to endorse it if it were functional.

**Comment by foolip** [comment]
This seems like a useful refactor, and I assume the intention is to not change any behavior. The very long list of composed helpers makes it clear that things are perhaps just a little bit convoluted. This refactor makes it hard to add additional state without it being obvious, so that's good, but I think ultimately we have some incidental suboptimal behavior, and fixing it necessarily means changing behavior. I would guess it's easier to change (and test) changes now however.

> One of my goals for this change is to produced machine readable logs from it. Likely an array of objects containing an UpdateState (or portion of it), the action taken (filter or persist), and reason for the action.

Yay to this goal 👍 

> I've also included a version of `inferSupportStatements`. It works in two loops, first to collapse ranges of null supported versions with their bookend versions, and the second turning that into `SimpleSupportStatement[]`.

This relates to the idea of validating consistency of (the absence of contradictions between) collector per-release results and BCD. Filling in null values is slightly at odds with that goal, because if BCD has some odd data for a range of releases, and we have no collector results for them, we should just leave them alone. In other words, I think a more feature complete update script probably has to do away with this behavior of `inferSupportStatements`, or at least keep track of what data is assumed and what is based on direct results.

**Comment by Elchi3** [comment]
@mzgoddard @foolip should we port this PR to https://github.com/openwebdocs/mdn-bcd-collector?
See also https://github.com/openwebdocs/mdn-bcd-collector/issues/257

**Review by limck620 — APPROVED** [review]

**Comment by foolip** [comment]
@Elchi3 I think that is the idea, yes. @mzgoddard can you close this PR and open a new one against the collector repo?

**Pull request closed without merging**